### PR TITLE
[SPARK-24350][SQL] Fixes ClassCastException in the "array_position" function

### PR DESCRIPTION
**Pull request by wajda**
### Fixes `ClassCastException` in the `array_position` function - [SPARK-24350](https://issues.apache.org/jira/browse/SPARK-24350)
When calling `array_position` function with a wrong type of the 1st argument an `AnalysisException` should be thrown instead of `ClassCastException`

Example:

```sql
select array_position('foo', 'bar')
```

```
java.lang.ClassCastException: org.apache.spark.sql.types.StringType$ cannot be cast to org.apache.spark.sql.types.ArrayType
	at org.apache.spark.sql.catalyst.expressions.ArrayPosition.inputTypes(collectionOperations.scala:1398)
	at org.apache.spark.sql.catalyst.expressions.ExpectsInputTypes$class.checkInputDataTypes(ExpectsInputTypes.scala:44)
	at org.apache.spark.sql.catalyst.expressions.ArrayPosition.checkInputDataTypes(collectionOperations.scala:1401)
	at org.apache.spark.sql.catalyst.expressions.Expression.resolved$lzycompute(Expression.scala:168)
	at org.apache.spark.sql.catalyst.expressions.Expression.resolved(Expression.scala:168)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveAliases$$anonfun$org$apache$spark$sql$catalyst$analysis$Analyzer$ResolveAliases$$assignAliases$1$$anonfun$apply$3.applyOrElse(Analyzer.scala:256)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveAliases$$anonfun$org$apache$spark$sql$catalyst$analysis$Analyzer$ResolveAliases$$assignAliases$1$$anonfun$apply$3.applyOrElse(Analyzer.scala:252)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:289)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:289)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:70)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:288)
```

## How was this patch tested?

unit test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 1]
```diff
@@ -708,6 +708,10 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       df.selectExpr("array_position(array(1, null), array(1, null)[0])"),
       Seq(Row(1L), Row(1L))
     )
+
+    intercept[AnalysisException] {
+      Seq(("a string element", "a")).toDF().selectExpr("array_position(_1, _2)")
+    }
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Please check the error message, too.

**Comment by maropu** [comment]
@ueshin can you trigger tests?

**Comment by ueshin** [comment]
ok to test.

**Comment by wajda** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by viirya** [comment]
`Fixes ClassCastException in the array_position function` - this seems like a better PR title than the current one.

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
good catch, thanks

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by viirya** [comment]
LGTM

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
LGTM

**Review by gatorsmile — COMMENTED** [review]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**